### PR TITLE
Merge xmr_btc crate

### PR DESCRIPTION
**Pull request by rishflab**
Created network, storage and protocol modules. Organised
files into the modules where the belong.

xmr_btc crate moved into isolated modulein swap crate.

Remove the xmr_btc module and integrate into swap crate.

Consolidate message related code

Reorganise imports

Remove unused parent Message enum

Remove unused parent State enum

Remove unused dependencies from Cargo.toml

### CODE REVIEW

**Comment by rishflab** [comment]
Retriggered CI due to failing Monero fund_transfer_and_check_tx_key test. Cannot reproduce locally.

**Review by da-kami — COMMENTED** [review]
I looked through the changes and am OK with merging this. I would recommend running it on `stagenet` as well before the merge to confirm that it works there too.
I just triggered a `stagenet` swap (alice and bob both on my machine). Will report back once it finishes.

File: swap/src/bitcoin.rs [hunk 1]
```diff
@@ -1,202 +1,288 @@
...
+use ecdsa_fun::{adaptor::Adaptor, fun::Point, nonce::Deterministic, ECDSA};
+use miniscript::{Descriptor, Segwitv0};
+use rand::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
+use sha2::Sha256;
+use std::str::FromStr;
+
+use crate::{config::Config, ExpiredTimelocks};
+
+use crate::bitcoin::timelocks::{BlockHeight, Timelock};
```
Comment (D4nte) [inline swap/src/bitcoin.rs]:
```suggestion
use std::str::FromStr;
use crate::{config::Config, ExpiredTimelocks};
use crate::bitcoin::timelocks::{BlockHeight, Timelock};
```

File: swap/src/bitcoin/transactions.rs [hunk 2]
```diff
@@ -12,6 +8,11 @@ use miniscript::{Descriptor, NullCtx};
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 
+use crate::bitcoin::{
+    build_shared_output_descriptor, timelocks::Timelock, verify_sig, BuildTxLockPsbt, Network,
```
Comment (D4nte) [inline swap/src/bitcoin/transactions.rs]:
```suggestion
use std::collections::HashMap;
use crate::bitcoin::{
    build_shared_output_descriptor, timelocks::Timelock, verify_sig, BuildTxLockPsbt, Network,
```

File: swap/src/bitcoin/wallet.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
+use ::bitcoin::{util::psbt::PartiallySignedTransaction, Address, Amount, Transaction, Txid};
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use backoff::{backoff::Constant as ConstantBackoff, future::FutureOperation as _};
+use bitcoin_harness::{bitcoind_rpc::PsbtBase64, BitcoindRpcApi};
+use reqwest::Url;
+use std::time::Duration;
+use tokio::time::interval;
+
+use crate::{
```
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:10]:
```suggestion
use tokio::time::interval;
use crate::{
```

File: swap/src/bitcoin/transactions.rs [hunk 4]
```diff
@@ -1,17 +1,18 @@
...
+    OutPoint, SigHash, SigHashType, TxIn, TxOut, Txid,
 };
+use anyhow::{bail, Context, Result};
 use ecdsa_fun::Signature;
 use miniscript::{Descriptor, NullCtx};
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 
+use crate::bitcoin::{
+    build_shared_output_descriptor, timelocks::Timelock, verify_sig, Address, Amount,
```
Comment (D4nte) [inline swap/src/bitcoin/transactions.rs:12]:
```suggestion
use std::collections::HashMap;
use crate::bitcoin::{
    build_shared_output_descriptor, timelocks::Timelock, verify_sig, Address, Amount,
```

File: swap/src/cli.rs [hunk 5]
```diff
@@ -2,6 +2,8 @@ use libp2p::{core::Multiaddr, PeerId};
 use url::Url;
 use uuid::Uuid;
 
+use crate::{bitcoin, monero};
```
Comment (D4nte) [inline swap/src/cli.rs:5]:
```suggestion
use url::Url;
use uuid::Uuid;
use crate::{bitcoin, monero};
```

**Comment by da-kami** [comment]
@rishflab the `xmr_btc` is actually not removed...
There is some leftover, including an empty `lib.rs` file when checking out the branch, see: https://github.com/comit-network/xmr-btc-swap/tree/reorganise-modules-2/xmr-btc/src

Shouldn't the complete crate be gone?

File: swap/src/database.rs [hunk 6]
```diff
@@ -3,13 +3,39 @@ use serde::{de::DeserializeOwned, Deserialize, Serialize};
 use std::{fmt::Display, path::Path};
 use uuid::Uuid;
 
-mod alice;
-mod bob;
+pub mod alice;
+pub mod bob;
```
Comment (D4nte) [inline swap/src/database.rs:7]:
```suggestion
mod alice;
mod bob;
```

Why is this needed? I created these modules purely to nicely organise the code within the `database` module, I did not intend them to be part of the `database` module API.

File: swap/src/database.rs [hunk 7]
```diff
@@ -3,13 +3,39 @@ use serde::{de::DeserializeOwned, Deserialize, Serialize};
...
+impl Display for Swap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Swap::Alice(alice) => Display::fmt(alice, f),
+            Swap::Bob(bob) => Display::fmt(bob, f),
+        }
+    }
+}
 
-#[derive(Debug)]
```
Comment (D4nte) [inline swap/src/database.rs:12]: Hum, did you *voluntarily* drop some clippy lints in the process?

File: swap/src/database/alice.rs [hunk 8]
```diff
@@ -1,14 +1,15 @@
-use crate::{alice::swap::AliceState, SwapAmounts};
-use bitcoin::hashes::core::fmt::Display;
+use ::bitcoin::hashes::core::fmt::Display;
 use serde::{Deserialize, Serialize};
-use xmr_btc::{
-    alice,
+
```
Comment (D4nte) [inline swap/src/database/alice.rs:3]:
```suggestion
```

File: swap/src/database/bob.rs [hunk 9]
```diff
@@ -1,7 +1,10 @@
-use crate::{bob::swap::BobState, SwapAmounts};
-use bitcoin::hashes::core::fmt::Display;
+use ::bitcoin::hashes::core::fmt::Display;
 use serde::{Deserialize, Serialize};
-use xmr_btc::bob;
+
```
Comment (D4nte) [inline swap/src/database/bob.rs:3]:
```suggestion
```

File: swap/src/lib.rs [hunk 10]
```diff
@@ -1,7 +1,5 @@
 #![warn(
     unused_extern_crates,
-    missing_debug_implementations,
```
Comment (D4nte) [inline swap/src/lib.rs:3]: Ok I see, ignore my comment asking about it.

File: swap/src/lib.rs [hunk 11]
```diff
@@ -1,7 +1,5 @@
...
-    missing_copy_implementations,
```
Comment (D4nte) [inline swap/src/lib.rs:4]: Thanks to your change I did a bit of reading: https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#missing-copy-implementations

I am happy eitherway, I think what is important is that if we decide to keep the lint then we must be weary to not put copy on data is likely to be of a huge size.

Consider the domain of this repo and the fact that we may have some huge byte array, removing the lint may actually be the better strategy 👍

File: swap/src/lib.rs [hunk 12]
```diff
@@ -10,19 +8,24 @@
     clippy::cast_possible_wrap,
     clippy::dbg_macro
 )]
+#![cfg_attr(not(test), warn(clippy::unwrap_used))]
 #![forbid(unsafe_code)]
-#![allow(non_snake_case)]
+#![allow(
+    non_snake_case,
+    missing_debug_implementations,
+    missing_copy_implementations
```
Comment (D4nte) [inline swap/src/lib.rs:16]:
```suggestion
```

These are allowed by default, no need to specify them.

File: swap/src/lib.rs [hunk 13]
```diff
@@ -1,7 +1,5 @@
 #![warn(
```
Comment (D4nte) [inline swap/src/lib.rs:1]: Be sure you have the exact same compiler attributes in `main.rs`

File: swap/src/lib.rs [hunk 14]
```diff
@@ -63,3 +66,10 @@ impl Display for SwapAmounts {
         )
     }
 }
+
+#[derive(Debug, Clone, Copy)]
+pub enum ExpiredTimelocks {
+    None,
+    Cancel,
+    Punish,
+}
```
Comment (D4nte) [inline swap/src/lib.rs:75]: This belongs in the `bitcoin` module. We want to remove `lib.rs` so let's not add more to it.

File: swap/src/main.rs [hunk 15]
```diff
@@ -250,9 +248,10 @@ async fn setup_wallets(
     bitcoin_wallet_name: &str,
     monero_wallet_rpc_url: url::Url,
     config: Config,
-) -> Result<(Arc<bitcoin::Wallet>, Arc<monero::Wallet>)> {
+) -> Result<(Arc<swap::bitcoin::Wallet>, Arc<swap::monero::Wallet>)> {
```
Comment (D4nte) [inline swap/src/main.rs:251]:
```suggestion
) -> Result<(Arc<bitcoin::Wallet>, Arc<monero::Wallet>)> {
```

Doesn't this work?

File: swap/src/monero.rs [hunk 16]
```diff
@@ -1,143 +1,377 @@
+pub mod wallet;
+
+use ::bitcoin::hashes::core::fmt::Formatter;
```
Comment (D4nte) [inline swap/src/monero.rs:3]:
```suggestion
```

This like a mistake where CLion imports `fmt` from `rust_bitcoin` instead of `std`

File: swap/src/monero.rs [hunk 17]
```diff
@@ -1,143 +1,377 @@
...
+    prelude::{FromPrimitive, ToPrimitive},
+    Decimal,
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    fmt::Display,
+    ops::{Add, Mul, Sub},
+    str::FromStr,
+};
+
```
Comment (D4nte) [inline swap/src/monero.rs:17]:
```suggestion
```

File: swap/src/monero/wallet.rs [hunk 18]
```diff
@@ -0,0 +1,147 @@
+use ::monero::{Address, Network, PrivateKey, PublicKey};
+use anyhow::Result;
+use async_trait::async_trait;
+use backoff::{backoff::Constant as ConstantBackoff, future::FutureOperation as _};
+use monero_harness::rpc::wallet;
+use std::{str::FromStr, time::Duration};
+use url::Url;
+
```
Comment (D4nte) [inline swap/src/monero/wallet.rs:8]:
```suggestion
```

File: swap/src/bitcoin.rs (hunk 1 above)
Comment (rishflab) [reply swap/src/bitcoin.rs]: hi, I don't understand this suggestion due to the code formatting. Do you want the `use crate::*` imports to NOT be separated from the `use *` imports with a line?

File: swap/src/protocol/alice.rs [hunk 19]
```diff
@@ -1,30 +1,32 @@
 //! Run an XMR/BTC swap in the role of Alice.
 //! Alice holds XMR and wishes receive BTC.
-use self::{amounts::*, message0::*, message1::*, message2::*, message3::*};
+pub use self::{amounts::*, message0::*, message1::*, message2::*, message3::*, state::*};
```
Comment (D4nte) [inline swap/src/protocol/alice.rs:3]: I usually prefer for the `pub use` to be separated so it's clear what does the module re-exports.

File: swap/src/lib.rs (hunk 13 above)
Comment (rishflab) [reply swap/src/lib.rs:1]: could remove the lib.rs now I think?

File: swap/src/lib.rs (hunk 11 above)
Comment (rishflab) [reply swap/src/lib.rs:4]: ok I will remove the missing copy lint

**Comment by da-kami** [comment]
> I looked through the changes and am OK with merging this. I would recommend running it on `stagenet` as well before the merge to confirm that it works there too.
> I just triggered a `stagenet` swap (alice and bob both on my machine). Will report back once it finishes.

Swapping on `stagenet` finished successfully for Alice and Bob.

File: swap/src/protocol/alice.rs [hunk 20]
```diff
@@ -133,10 +135,10 @@ impl From<message3::OutEvent> for OutEvent {
...
     pt: PeerTracker,
     amounts: Amounts,
-    message0: Message0,
-    message1: Message1,
-    message2: Message2,
-    message3: Message3,
+    message0: Message0Behaviour,
+    message1: Message1Behaviour,
+    message2: Message2Behaviour,
+    message3: Message3Behaviour,
```
Comment (D4nte) [inline swap/src/protocol/alice.rs:141]:
```suggestion
    message0: message0::Behaviour,
    message1: message1::Behaviour,
    message2: message2::Behaviour,
    message3: message3::Behaviour,
```

I believe it is better to use the module to qualify a type than having (redundant because right now it is `message0::Message0Behaviour`) qualifier in the name.

File: swap/src/protocol/bob.rs [hunk 21]
```diff
@@ -1,24 +1,22 @@
...
-use self::{amounts::*, message0::*, message1::*, message2::*, message3::*};
-use crate::{
-    network::{
-        peer_tracker::{self, PeerTracker},
-        transport::SwapTransport,
-        TokioExecutor,
-    },
-    SwapAmounts,
-};
+pub use self::{amounts::*, message0::*, message1::*, message2::*, message3::*, state::*};
```
Comment (D4nte) [inline swap/src/protocol/bob.rs:3]: So `messageN` exports a `MessageNBehaviour`, `MessageN` and `OutEvent`.

I think it would be cleaner to:
- rename the types to `Message` (or event `Content` or `Body`) and `Behaviour`
- Not do a wildcard re-export (remove `pub use`)
- Use `message0::Behaviour` and `message0::Content` wherever needed.

File: swap/src/protocol/bob/message1.rs [hunk 22]
```diff
@@ -6,15 +6,25 @@ use libp2p::{
     swarm::{NetworkBehaviourAction, NetworkBehaviourEventProcess, PollParameters},
     NetworkBehaviour, PeerId,
 };
+use serde::{Deserialize, Serialize};
 use std::{
     collections::VecDeque,
     task::{Context, Poll},
     time::Duration,
 };
+
```
Comment (D4nte) [inline swap/src/protocol/bob/message1.rs:15]:
```suggestion
```

This new line is actually not needed.

File: swap/src/protocol/bob/message2.rs [hunk 23]
```diff
@@ -6,15 +7,25 @@ use libp2p::{
     swarm::{NetworkBehaviourAction, NetworkBehaviourEventProcess, PollParameters},
     NetworkBehaviour, PeerId,
 };
+use serde::{Deserialize, Serialize};
 use std::{
     collections::VecDeque,
     task::{Context, Poll},
     time::Duration,
 };
+
```
Comment (D4nte) [inline swap/src/protocol/bob/message2.rs:16]:
```suggestion
```

This is actually not needed even if we keep the segregation of `crate` imports.

**Review by D4nte — APPROVED** [review]

File: swap/src/bitcoin.rs (hunk 1 above)
Comment (D4nte) [reply swap/src/bitcoin.rs]: I don't mind the segregation of the `use crate::*`.

Do note that in the original code there is a newline between both `use crate::*`, that needs to be removed.

File: swap/src/lib.rs (hunk 11 above)
Comment (D4nte) [reply swap/src/lib.rs:4]: > ok I will remove the missing copy lint

It is already removed, no action is necessary for this comment.

File: swap/src/lib.rs (hunk 13 above)
Comment (D4nte) [reply swap/src/lib.rs:1]: After reviewing, I'd suggest to not worry about the lints and lib.rs within the scope of this PR.

File: swap/src/protocol.rs [hunk 24]
```diff
@@ -1,2 +1,49 @@
...
+// pub enum Cmd {
+//     VerifyAmounts(SwapAmounts),
+// }
+//
+// /// Responses sent from the main task back to Bob.
+// #[derive(Clone, Copy, Debug, PartialEq)]
+// pub enum Rsp {
+//     VerifiedAmounts,
+//     Abort,
+// }
```
Comment (D4nte) [inline swap/src/protocol.rs:49]:
```suggestion
```

File: swap/src/tests/happy_path.rs [hunk 25]
```diff
@@ -1,22 +1,22 @@
...
     future::{join, select},
     FutureExt,
 };
 use get_port::get_port;
 use libp2p::Multiaddr;
 use rand::rngs::OsRng;
-use swap::{
+use testcontainers::clients::Cli;
+
+use uuid::Uuid;
```
Comment (D4nte) [inline swap/src/tests/happy_path.rs]:
```suggestion
use testcontainers::clients::Cli;
use uuid::Uuid;
```

File: swap/src/main.rs [hunk 26]
```diff
@@ -14,25 +14,31 @@
...
 use uuid::Uuid;
 
+pub mod bitcoin;
+pub mod cli;
+pub mod config;
+pub mod database;
+pub mod monero;
+pub mod network;
+pub mod protocol;
+#[cfg(test)]
```
Comment (D4nte) [inline swap/src/main.rs:38]: That's what I am talking about :)

Excellent, now I can go and isolate our `run_until` function that only exists for test purposes :)

**Review by D4nte — APPROVED** [review]

File: swap/src/protocol/bob.rs (hunk 21 above)
Comment (rishflab) [reply swap/src/protocol/bob.rs:3]: hmm I didn't `messageN::MessageN` to `messageN::Content` because it used in a lot in the rest of the codebase and I wanted to avoid having `alice::messageN::Content` everywhere.

File: swap/src/database.rs (hunk 6 above)
Comment (rishflab) [reply swap/src/database.rs]: its not, I had them Alice and bob in the same file till things got messy with a rebase.

File: swap/src/database.rs (hunk 7 above)
Comment (rishflab) [reply swap/src/database.rs:12]: Fixed this in a follow up commit, not sure what happened here tbh

File: swap/src/main.rs (hunk 26 above)
Comment (rishflab) [reply swap/src/main.rs:38]: we are calling `run_until()` in `swap()`. How would you remove `run_until()` from production code without causing code duplication?

Comment (D4nte) [reply swap/src/main.rs:38]: > we are calling run_until() in swap(). How would you remove run_until() from production code without causing code duplication?

I am happy to take over this change myself at a later stage.
